crates/wiggle/src/guest_type.rs

```diff
@@ -88,7 +91,19 @@ macro_rules! primitives {
                 if ptr.mem().is_mut_borrowed(region) {
                     return Err(GuestError::PtrBorrowed(region));
                 }
-                Ok(unsafe { <$i>::from_le_bytes(*host_ptr.cast::<[u8; mem::size_of::<Self>()]>()) })
+                // If the accessed memory is shared, we need to load the bytes
+                // with the correct memory consistency.
+                let value = if ptr.mem().is_shared_memory() {
+                    let atomic_value_ref: &$AtomicI = unsafe { &*(host_ptr as *mut $i as *const $AtomicI) };
+                    let value = atomic_value_ref.load(Ordering::Relaxed);
+                    // For floating-point values, we must transmute from their
+                    // integer equivalent (see
+                    // https://doc.rust-lang.org/std/primitive.f32.html#method.to_bits).
+                    unsafe { std::mem::transmute::<_, $i>(value) }
```

Comment:
I think that this'll need the same `*_le_*` treatment as below, although I realize that'll get funky with the macro here.

For converting between `f32` and `i32` could `f32::from_bits` and `to_bits` be used instead of `transmute` as well?